ironfish-rust/src/public_account/description.rs

```diff
@@ -0,0 +1,275 @@
+            signers.push(signer);
+        }
+
+        let mut signatures_len_buf = [0; 2];
+        reader.read_exact(&mut signatures_len_buf)?;
+        let signatures_len = u16::from_le_bytes(signatures_len_buf) as usize;
+
+        let mut signatures = Vec::with_capacity(signatures_len);
+        for _ in 0..signatures_len {
+            let mut signature_buf = [0; 64];
+            reader.read_exact(&mut signature_buf)?;
+            let signature = Signature::from_bytes(&signature_buf);
+            signatures.push(signature);
+        }
+
+        let mut transfers_len_buf = [0; 2];
+        reader.read_exact(&mut transfers_len_buf)?;
+        let transfers_len = u16::from_le_bytes(transfers_len_buf) as usize;
+
+        let mut transfers = Vec::with_capacity(transfers_len);
+        for _ in 0..transfers_len {
+            let transfer = Transfer::read(&mut reader)?;
+            transfers.push(transfer);
+        }
+        let address = PublicAddress::read(&mut reader)?;
+
+        let description = Self {
+            version,
+            min_signers,
+            signers,
+            signatures,
+            address,
+            transfers,
+        };
+        description.valid()?;
+        Ok(description)
+    }
+
+    pub fn write<W: io::Write>(&self, mut writer: W) -> Result<(), IronfishError> {
+        // TODO: think about ideal ordering here
+        writer.write_all(&self.version.to_le_bytes())?;
+
+        writer.write_all(&self.min_signers.to_le_bytes())?;
+
+        let signers_len: u16 = self.signers.len().try_into()?;
+        writer.write_all(&signers_len.to_le_bytes())?;
+
+        for signer in &self.signers {
+            writer.write_all(signer.as_bytes())?;
+        }
+
+        let signatures_len: u16 = self.signatures.len().try_into()?;
+        writer.write_all(&signatures_len.to_le_bytes())?;
+
+        for signature in &self.signatures {
+            writer.write_all(&signature.to_bytes())?;
+        }
+
+        let transfers_len: u16 = self.transfers.len().try_into()?;
+        writer.write_all(&transfers_len.to_le_bytes())?;
+
+        for transfer in &self.transfers {
+            transfer.write(&mut writer)?;
+        }
+
+        self.address.write(&mut writer)?;
+
+        Ok(())
+    }
+
+    fn valid(&self) -> Result<(), IronfishError> {
+        if self.min_signers < 1 {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidThreshold));
+        }
+
+        if self.signers.len() < self.min_signers as usize {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidThreshold));
+        }
+
+        let unique_signers: Vec<_> = self.signers.iter().collect();
+        if unique_signers.len() != self.signers.len() {
+            return Err(IronfishError::new(IronfishErrorKind::DuplicateSigner));
+        }
+
+        let unique_signatures: Vec<_> = self.signatures.iter().collect();
+        if unique_signatures.len() != self.signatures.len() {
```

Comment:
Note that this works only because Ed25519 signatures are deterministic. If we were using a different signature scheme (e.g. ECDSA), then this wouldn't guarantee that a single signer didn't add multiple signatures, which would make the `min_signatures` test pass even though the threshold wasn't met.

To prevent this kind of problems, I would count the number of signers that contributed a signature instead (here or in the `verify` method):

```rust
let signers = HashMap::new();
for signature in &self.signatures {
    let signer = self
        .signers
        .iter()
        .filter(|signer| signer.verify(...))
        .next();
    match signer {
        None => return Err(...),
        Some(signer) => signers.entry(signer.clone()).and_modify(|count| count += 1).or_insert(1),
    }
}
```